### ISSUE TITLE
emscripten::val fails in global scope constructors

### ISSUE THREAD

**Issue by slowriot** [emscripten-core/emscripten#23170]
Tested with 3.1.73.

Attempting to use `val` in constructors executed at global scope (before `main` begins) fails with `emval::as has unknown type` errors.

Minimal example:

```cpp
#include <iostream>
#include <emscripten/emscripten.h>
#include <emscripten/val.h>

struct test_struct {
  test_struct() {
    std::cout << "nagivator.userAgent " << emscripten::val::global("navigator")["userAgent"].as<std::string>() << std::endl;
    std::cout << "window.innerWidth " << emscripten::val::global("window")["innerWidth"].as<unsigned int>() << std::endl;
  }
};

// fails when constructed in global scope:
test_struct test;

auto main()->int {
  // OK :
  //test_struct test;

  // also OK:
  //static test_struct test;

  std::cout << "Hello world" << std::endl;
  return EXIT_SUCCESS;
}
```
Build with `--bind` and `-sENVIRONMENT=web`.

Expected output, and output when test_struct is initialised inside `main`:
![image](https://github.com/user-attachments/assets/774d7be7-68da-4561-994a-ebdd2af0142a)

Output when trying to initialise at global scope, for the `.as<std::string>()`:
![image](https://github.com/user-attachments/assets/cd04c0de-0def-493c-a2c4-d70799e87cee)
and for the `.as<unsigned int>()`:
![image](https://github.com/user-attachments/assets/3b92a535-7c3c-46e6-9cf8-8c35415749c3)

I'm guessing that something gets initialised before `main` runs, but after static construction completes.  Whereas ideally, everything should be ready before any C++ objects are constructed at static scope too.

**Comment by sbc100** [comment]
Hmm, IIRC we do all the embind type registration because user-level constructors are called, but I looks like the ordering is wrong here. 

Actually looking at the code it looks like we do use a C++ constructor to register the types: https://github.com/emscripten-core/emscripten/blob/dfe3356cb1860ebf087de33f85a742e8669cce5b/system/include/emscripten/bind.h#L2267-L2280.  So because ordering of C++ constructors between object files is not something we can control I think that means the emval is not currently usable in C++ constructors (is that right @brendandahl ?)

**Comment by slowriot** [comment]
To be clear, `val` works fine in the constructor as long as `test_struct` is initialised in `main()` or in objects initialised after `main` begins.  It only fails for objects with static storage duration at global scope.  Function-scope statics are fine too, presumably because they initialise after main() begins.  So it's definitely not an issue for *all* constructors.

**Comment by sbc100** [comment]
> So it's definitely not an issue for _all_ constructors.

Sorry I mean "static constructor" in the sense of "constructors that run before main runs".    The reason is that embind is itself initialized by such a "static constructor", and you have no way of ensuring that it happens before you one does.   You could get lucky, or try to play with object file ordering on the command line.. but that super fragile.

**Comment by brendandahl** [comment]
> So because ordering of C++ constructors between object files is not something we can control I think that means the emval is not currently usable in C++ constructors (is that right @brendandahl ?)

Yeah, it's best to avoid using emval during static constructors. We should probably add a note in the documentation or I suppose we could add an assert.

**Comment by sbc100** [comment]
We used to use a C static constructor to try to work around this, but it turned out that had its own issues.  See #17182.

**Comment by msorvig** [comment]
We could possibly fix this by using init_priority for the builtins in bind.cpp, only. That way those gets initialized first, and we avoid changing the inti ordering for user EMSCRIPTEN_BINDINGS, which was the bug in #17182.
